### DISCUSSION TITLE
Support `releaseTimestamp` in the docker datasource

### DISCUSSION THREAD

**Discussion by pathob** [renovatebot/renovate#26180] [Suggest an Idea]
### Tell us more.

The use / provision of the `releaseTimestamp` attribute is not consistent yet among all datasources. For example it is available for Maven and Helm, but not for Docker.

This discussion aims for having a feature request to support `releaseTimestamp` in the docker datasource

@rarkins following discussion #24714

**Comment by viceice** [comment]
only possible on docker hub with `RENOVATE_X_DOCKER_HUB_TAGS=true` as self-hosted user. Otherwise not possible because of docker api limitations 🤷‍♂️

**Reply by pathob** [reply]
I see, but that reminds me of another discussion with @rarkins about the `appVersion` in Helm charts (https://github.com/renovatebot/renovate/discussions/20017#discussioncomment-7754202), that is not available in the `index.yml` of a Helm repository, but only in the `.tar.gz`'s `Chart.yml`.

If the image would be downloaded, we could use `docker image inspect` to get the `Created` attribute.

Maybe Renovate should be able to support these cases?

At least, what both cases have in common, is, that the artifact would only needed to be downloaded if Renovate detected a potential upgrade already.

**Reply by viceice** [reply]
That will be very expensive if we do that for every version, so a nogo.

**Reply by pathob** [reply]
Not per default, but optionally

**Reply by pathob** [reply]
For my client the `releaseTimestamp`'s are very important. He would definitely accept these expensive Renovate operations to get this information.

**Reply by viceice** [reply]
if you do this against docker hub than you'll get rate limited very fast. so it won't be enabled on official renovate apps.

is your client using docker hub? than he should use `RENOVATE_X_DOCKER_HUB_TAGS=true` which will be enabled by default soon. It will fetch docker tag from the hub api which provides the `releaseTimestamp`.

**Reply by pathob** [reply]
No, they are proxying via Artifactory (and they have a Dockerhub subscription for not running into rate limits). The images are also only partially coming from Dockerhub, but also from GHCR, Quay, etc.

**Comment by rarkins** [comment]
For Docker Hub images (i.e. library images), set `RENOVATE_X_DOCKER_HUB_TAGS=true` in env as described by @viceice below.

For other Docker images, we should look into the docker protocol/API to work out how a timestamp can be retrieved per tag/digest as efficiently as possible. If we need to pull the the entire image then that's about as inefficient as you can get, and hopefully there's a better way. If we keep  the digest -> releaseTimestamp mapping cached _permanently_ and only do this "lazily" when it's part of a PR, then maybe it's feasible.

**Reply by viceice** [reply]
the most efficient way is to use a `HEAD` request on the manifest (which we do to get the digest), but still  one request per version (tags are mutable).

**Reply by rarkins** [reply]
There seems to be quite a lot of questions like this even in the 2015-2020 timeframe, e.g. https://stackoverflow.com/questions/60059226/how-to-get-the-last-pushed-date-of-a-docker-image

@pathob Artifactory often does the "least possible" protocol compliance though so you'd definitely want to test it manually locally to make sure they include the same metadata as we would need.

BTW we also need to consider some complexity:
- A tag, such as `node:18`, can be pushed to many times over its life so therefore its timestamp should reflect the last time it's pushed to
- A tag usually has multiple digests, e.g. per-architecture and also a general one representing all, and each might have their own time stamp as they can be published somewhat independently

**Reply by viceice** [reply]
@rarkins This should be optional and disabled by default to not overload registries.

**Reply by pathob** [reply]
> @pathob Artifactory often does the "least possible" protocol compliance though so you'd definitely want to test it manually locally to make sure they include the same metadata as we would need.

I will have a look at it beginning of next week.

> @rarkins This should be optional and disabled by default to not overload registries.

I'm not the mentioned one, but I agree.

Thanks for the discussion!

**Reply by pathob** [reply]
Okay, so I was mistaken, they are now using Harbor, not Artifactory anymore. But anyway.

No news here. The API used in the stackoverflow question is not standard, Artifactory and Harbor don't have them. Docker image manifests don't contain the information, so only `docker image inspect` requiring full image download provides the information...

**Comment by PG2000** [comment]
Would it make sense to have a strategy to implement this for other Registries as well?

For example in Harbor the URL would look like this?

```bash
curl -X 'GET' \
  'https://private-registry.example.local/api/v2.0/projects/helm-charts/repositories/bitnami%2Fexternal-dns/artifacts/6.30.0/tags?page=1&page_size=10' \
  -H 'accept: application/json'
```

and the result would contain the `push_time`

```json
[
  {
    "artifact_id": 28114,
    "id": 11472,
    "immutable": false,
    "name": "6.30.0",
    "pull_time": "2024-01-19T04:03:39.665Z",
    "push_time": "2024-01-17T12:09:07.790Z",
    "repository_id": 529
  }
]
```

**Reply by viceice** [reply]
is harbor detectable via a http header when calling `https://some.domain.test/v2/` ?

**Reply by PG2000** [reply]
Not from my knowledge. A quick test had the response `x-harbor-csrf-token` but i think this is not reliable enough as the use case is a different one for this header.
In my case it would be possible to add some extra response header.

Do you have a solution in mind how this could be solved if Harbor could be detected via Header?

**Reply by viceice** [reply]
yes, we already do that mentioned request to check if auth is required on first request to a docker registry. we could use that to detect harbor. we do something similar for artifactory already.
